File: 0051-n-queens/0051-n-queens.cpp
class Solution {
public:
    vector<vector<string>> answer;
    set<int> antiDiagonals;
    set<int> diagonals;
    set<int> verticals;

    void backtrack(int row, int n, vector<string>& board) {
        if (row == n) {
            answer.push_back(board);
            return;
        }

        for (int col=0; col<n; col++) {
            if (antiDiagonals.count(row+col) || diagonals.count(row-col)
                || verticals.count(col)) {
                    continue;
                }

            antiDiagonals.insert(row+col);
            diagonals.insert(row-col);
            verticals.insert(col);
            board[row][col] = 'Q';
            
            backtrack(row+1, n, board);

            board[row][col] = '.';
            verticals.erase(col);
            diagonals.erase(row-col);
            antiDiagonals.erase(row+col);
        }

        return;
    }

    vector<vector<string>> solveNQueens(int n) {
        vector<string> board;
        string dots;

        for (int i=0; i<n; i++) {
            dots += ".";
        }
        for (int i=0; i<n; i++) {
            board.push_back(dots);
        }

        backtrack(0, n, board);
        return answer;
    }
};

File: 0051-n-queens/0051-n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        output = []
        board = ["."*n] * n
        verticals = [False] * n 

        def checkDiagonal(board, x, y):
            i, j = x+1, y+1
            while i < n and j < n:
                if board[i][j] == "Q":
                    return False
                i += 1
                j += 1

            i, j = x-1, y+1
            while i >= 0 and j < n:
                if board[i][j] == "Q":
                    return False
                i -= 1
                j += 1    

            i, j = x-1, y-1
            while i >= 0 and j >= 0:
                if board[i][j] == "Q":
                    return False
                i -= 1
                j -= 1

            i, j = x+1, y-1
            while i < n and j >= 0:
                if board[i][j] == "Q":
                    return False
                i += 1
                j -= 1

            return True            


        def dfs(board, x, y):
            if x == n:
                output.append(board[:])
                return

            for y in range(n):
                if verticals[y]:
                    continue
                if not checkDiagonal(board, x, y):
                    continue
                
                temp = ["."]*n
                temp[y] = "Q"
                board[x] = "".join(temp)
                verticals[y] = True

                dfs(board, x+1, 0)

                verticals[y] = False
                board[x] = "." * n

        dfs(board, 0, 0)

        return output